File: homework_1/hermitcrabs.py
def insertionsort(values, n):
    # running in O(n^2) time
    for i in range(n): 
        val = values[i] # get the current value being looked at to be sorted in
        j = i - 1 # start at value directly left of current value
        # loop throgh elements from 0 to i-1
        while(values[j] >= val and j >= 0):
            # if the left value is greater then the right value, swap with vals position
            values[j + 1] = values[j]
            values[j] = val # swap it with j
            j -= 1 # decrease j
    return values

def merge(arr1, arr2):
    output = [] #output array is len(arr1) + len(arr2) size
    i, j, k = 0, 0, 0 # initialize variables
    while (i < len(arr1) and j < len(arr2)): #while one of the arrays is not empty
        
        # add first arrays smaller value
        if (arr1[i] <= arr2[j]):
            output.append(arr1[i])
            i += 1

        # add second arrays smaller value
        else:
            output.append(arr2[j])
            j += 1
        k += 1 # k is not actually needed to be incremented but in the pseudocode from the slides it was used, since list don;t need predefined size like array in java, it isnt needed
    
    #append the rest of the remaining array
    if i < len(arr1):
        output.extend(arr1[i:])
    elif j < len(arr2):
        output.extend(arr2[j:])
    return output


def mergesort(values, n):
    #trying insertion sort for slightly larger base case to speed things up to meet time limit
    if n<=20: # base case, it is length of 1
        return insertionsort(values, n)
    else:
        mid = n//2  # get middle index
        arr1 = mergesort(values[:mid], mid) # call merge sort on left
        arr2 = mergesort(values[mid:], n - mid) # call mergesort on right
        values = merge(arr1, arr2) # merge them back together
        return values
    

def main():

    # parse the input

    # Get first two integers n and m
    first = input().strip().split()
    n = int(first[0])
    m = int(first[1])

    # get n integers
    crabs = mergesort([int(x) for x in input().strip().split()], n) # get the crabs and sort them

    # m integers of empty shell sizes
    shells = mergesort([int(x) for x in input().strip().split()], m) #get the empty shells and sort them

    c = n - 1  # index for the largest crab to fit
    s = m - 1 # index the largest shell
    
    
    while c > 0 and s > 0: # while there are still crabs left to assign and possible shells to assign
        if shells[s] > crabs[c]: #if current shell is bigger than current crabs size
            # assign this shell to this crab
            shells.pop() # remove the shell just used
            shells.append(crabs[c]) # add the emptied shell to end
            c -= 1 # go to next smaller crab
            # s = len(shells) - 1 # point shell pointer to the last shell
        else:
            # shell too small--> try next shell
            s -= 1

    print("YES" if c == 0 else "NO")

if __name__ == "__main__":
    main()


    #need to add leftshell back to the list

File: homework_1/src/java/main/HermitCrabs.java
import java.util.Scanner;

/**
 * HermitCrabs.java
 * I had to write this in java because the python version was timing out for the last two, so
 * that is why this is in java and the rest of the assignments are in Python...
 */
public class HermitCrabs {

    /**
     * Merges the arrays together (mergesort heloer).
     *
     * @param arr1 first sorted array
     * @param arr2 second sorted array
     * @return cmerged array
     */
    static int[] merge(int[] arr1, int[] arr2) {

        //merged arr has length of the two combined
        int[] merged_arr = new int[arr1.length + arr2.length];
        int i = 0, j = 0, k = 0; //insitialize variuables

        //while neither arrays are empty
        while (i < arr1.length && j < arr2.length) {
            //if the element in first array is smaller, add that
            if (arr1[i] <= arr2[j]) merged_arr[k++] = arr1[i++];
            else merged_arr[k++] = arr2[j++]; //o.w. add the second array elemnt
        }
        
        //add the rest of the non-empty array to the merged_Array at the end
        while (i < arr1.length){ merged_arr[k++] = arr1[i++];}
        while (j < arr2.length) {merged_arr[k++] = arr2[j++];}

        return merged_arr;
    }

    /**
     * mergesort sorts and array in nlogn time
     *
     * @param values array to sort
     * @return sorted array
     */
    static int[] mergeSort(int[] values) {
        int n = values.length; //get length of values to sort

        //base case, if it is length 1 then return
        if (n <= 1) {
            return values;
        }

        //split it up
        int mid = n / 2;

        //get new array of proper lengths
        int[] left = new int[mid];
        int[] right = new int[n - mid];

        // copy the first and seocnd half into the right array
        for (int i = 0; i < mid; i++) left[i] = values[i];
        for (int i = mid; i < n; i++) right[i - mid] = values[i];

        //call merge sort on each side
        left = mergeSort(left);
        right = mergeSort(right);

        //merger together
        return merge(left, right);
    }

    /**
     * main program for hermit crabs
     */
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        //scan in the number of crabs and shells
        int n = sc.nextInt();
        int m = sc.nextInt();

        //get the crabs list and sort it
        int[] crabs = new int[n];
        for (int i = 0; i < n; i++) crabs[i] = sc.nextInt();
        crabs = mergeSort(crabs);

        //get the shells list and sort it
        int[] shells = new int[m];
        for (int i = 0; i < m; i++) shells[i] = sc.nextInt();
        shells = mergeSort(shells);

        //get starting idx for each list
        int c = n - 1;
        int s = m - 1;

        //while not at the start of the array
        while (c > 0 && s > 0) { 
            //if shell is bigger
            if (shells[s] > crabs[c]) {
                // overwrite last element with the new shell
                shells[m - 1] = crabs[c];
                c--; //go to smaller crab
                // s stays the same because now pointing to next open shell
            } else {
                s--; //go to next shell sinice it doesnt fit :()
            }
        }

        //return output
        System.out.println(c == 0 ? "YES" : "NO");
        sc.close();
    }
}
